File: api/Domain/Model/PhysicalResources/PhysicalResource.cs
using Api.Domain.Model;
using Domain.Model.Generic;

public enum ResourceStatus
{
    Available = 0,
    Maintenance = 1,
    OutOfService = 2
}

// https://learn.microsoft.com/en-us/ef/core/modeling/inheritance
public class PhysicalResource : IDTOAble<PhysicalResourceDto>
{
    public Guid Id { get; private set; }
    public Code Code { get; private set; }
    public Designation Description { get; private set; }
    public ResourceStatus Status { get; private set; }
    public TimeSpan SetupTime { get; private set; } // In minutes
    public OperationalWindow OperationalWindow { get; private set; }
    public bool Active { get; private set; } = true; // Soft delete
    public virtual ICollection<Qualification> Qualifications { get; private set; } // Needed qualifications to operate the resource

    protected PhysicalResource() { } // EF Core

    internal PhysicalResource(Guid id, Code code, Designation description, ResourceStatus status, TimeSpan setupTime, HashSet<Qualification> qualifications, OperationalWindow operationalWindow)
    {
        Id = id;
        Description = description;
        Status = status;
        SetupTime = setupTime;
        Code = code;
        Qualifications = qualifications;
        OperationalWindow = operationalWindow;
    }
    
    public void Deactivate() { Active = false; }

    public void UpdateDescription(Designation description) { Description = description; }
    public void UpdateStatus(ResourceStatus status) { Status = status; }
    public void UpdateSetupTime(TimeSpan setupTime) { SetupTime = setupTime; }
    public void UpdateQualifications(HashSet<Qualification> qualifications) { Qualifications = qualifications; }
    public void UpdateOperationalWindow(OperationalWindow operationalWindow) { OperationalWindow = operationalWindow; }

    public PhysicalResourceDto ToDTO()
    {
        return new PhysicalResourceDto
        {
            Code = this.Code.ToString(),
            Description = this.Description.ToString(),
            Status = this.Status,
            SetupTimeInMinutes = (int)this.SetupTime.TotalMinutes,
            Qualifications = this.Qualifications.Select(q => q.ToDTO()).ToList(),
            OperationalWindow = this.OperationalWindow
        };
    }
}

public class STSCrane : PhysicalResource, IDTOAble<STSCraneDto>
{
    public uint ContainersPerHour { get; private set; }
    public uint LiftingCapacity { get; private set; }
    public virtual Dock ServingDock { get; private set; }

    protected STSCrane() { } // EF Core

    public STSCrane(
        Guid id, Code code, Designation description, ResourceStatus status, TimeSpan setupTime, HashSet<Qualification> qualifications, OperationalWindow operationalWindow,
        uint liftingCapacity, Dock dock, uint averageContainersPerHour
    )
        : base(id, code, description, status, setupTime, qualifications, operationalWindow)
    {
        LiftingCapacity = liftingCapacity;
        ServingDock = dock;
        ContainersPerHour = averageContainersPerHour;
    }

    public void UpdateServingDock(Dock dock) { ServingDock = dock; }
    public void UpdateLiftingCapacity(uint liftingCapacity) { LiftingCapacity = liftingCapacity; }
    public void UpdateContainersPerHour(uint containersPerHour) { ContainersPerHour = containersPerHour; }

    STSCraneDto IDTOAble<STSCraneDto>.ToDTO()
    {
        return new STSCraneDto
        {
            Code = this.Code.ToString(),
            Description = this.Description.ToString(),
            Status = this.Status,
            SetupTimeInMinutes = (int)this.SetupTime.TotalMinutes,
            Qualifications = this.Qualifications.Select(q => q.ToDTO()).ToList(),
            ContainersPerHour = this.ContainersPerHour,
            LiftingCapacity = this.LiftingCapacity,
            ServingDock = this.ServingDock.ToDTO(),
            OperationalWindow = this.OperationalWindow
        };
    }
}

public class YardCrane : PhysicalResource, IDTOAble<YardCraneDto>
{
    public uint ContainersPerHour { get; private set; }
    public uint LiftingCapacity { get; private set; }
    private StorageArea _yardSection;
    public virtual StorageArea YardSection
    {
        get => _yardSection;
        private set
        {
            if (value.AreaType != StorageAreaType.Yard)
                throw new ArgumentException("The storage area must be of type 'Yard'.");

            _yardSection = value;
        }
    }

    protected YardCrane() { } // EF Core

    public YardCrane(
        Guid id, Code code, Designation description, ResourceStatus status, TimeSpan setupTime, HashSet<Qualification> qualifications, OperationalWindow operationalWindow,
        uint liftingCapacity, StorageArea yardSection, uint averageContainersPerHour
    )
        : base(id, code, description, status, setupTime, qualifications, operationalWindow)
    {
        LiftingCapacity = liftingCapacity;
        YardSection = yardSection;
        ContainersPerHour = averageContainersPerHour;
    }

    public void UpdateYardSection(StorageArea yardSection) { YardSection = yardSection; }
    public void UpdateLiftingCapacity(uint liftingCapacity) { LiftingCapacity = liftingCapacity; }
    public void UpdateContainersPerHour(uint containersPerHour) { ContainersPerHour = containersPerHour; }
    
    YardCraneDto IDTOAble<YardCraneDto>.ToDTO()
    {
        return new YardCraneDto
        {
            Code = this.Code.ToString(),
            Description = this.Description.ToString(),
            Status = this.Status,
            SetupTimeInMinutes = (int)this.SetupTime.TotalMinutes,
            Qualifications = this.Qualifications.Select(q => q.ToDTO()).ToList(),
            ContainersPerHour = this.ContainersPerHour,
            LiftingCapacity = this.LiftingCapacity,
            YardSection = this.YardSection.ToDTO(),
            OperationalWindow = this.OperationalWindow
        };
    }
}


public class Truck : PhysicalResource, IDTOAble<TruckDto>
{
    public uint ContainersPerTrip { get; private set; } // Number of containers the truck can carry per trip
    public uint AverageSpeed { get; private set; } // In km/h
    public uint MaxLoadCapacity { get; private set; } // In kg

    protected Truck() { } // EF Core

    public Truck(
        Guid id, Code code, Designation description, ResourceStatus status, TimeSpan setupTime, HashSet<Qualification> qualifications, OperationalWindow operationalWindow,
        uint maxLoadCapacity, uint containersPerTrip, uint averageSpeed
    )
        : base(id, code, description, status, setupTime, qualifications, operationalWindow)
    {
        MaxLoadCapacity = maxLoadCapacity;
        ContainersPerTrip = containersPerTrip;
        AverageSpeed = averageSpeed;
    }

    public void UpdateContainersPerTrip(uint containersPerTrip) { ContainersPerTrip = containersPerTrip; }
    public void UpdateAverageSpeed(uint averageSpeed) { AverageSpeed = averageSpeed; }
    public void UpdateMaxLoadCapacity(uint maxLoadCapacity) { MaxLoadCapacity = maxLoadCapacity; }

    TruckDto IDTOAble<TruckDto>.ToDTO()
    {
        return new TruckDto
        {
            Code = this.Code.ToString(),
            Description = this.Description.ToString(),
            Status = this.Status,
            SetupTimeInMinutes = (int)this.SetupTime.TotalMinutes,
            Qualifications = this.Qualifications.Select(q => q.ToDTO()).ToList(),
            ContainersPerTrip = this.ContainersPerTrip,
            MaxLoadCapacity = this.MaxLoadCapacity,
            AverageSpeed = this.AverageSpeed,
            OperationalWindow = this.OperationalWindow
        };
    }
}

File: FrontEnd/src/model/PhysicalResource.ts
import type { Dock } from "./Dock";
import type { Qualification } from "./Qualifications";
import type { OperationalWindow } from "./values/OperationalWindow";
import type { PhysicalResourceDto, STSCraneDto, YardCraneDto, TruckDto } from "./dto/PhysicalResourceDto";

export interface PhysicalResourceFilter {
    Code?: string;
    Description?: string;
    Status?: 0 | 1 | 2;
    Type?: 0 | 1 | 2;
}

export enum ResourceStatus {
    Available = 0,
    Maintenance = 1,
    OutOfService = 2
}

export abstract class PhysicalResource {
    protected _type: string;
    protected _code: string;
    protected _description: string;
    protected _status: ResourceStatus;
    protected _setupTimeInMinutes: number;
    protected _operationalWindow: OperationalWindow;
    protected _qualifications: Qualification[];
    protected _active: boolean = true;

    constructor(params: {
        type: string;
        code: string;
        description: string;
        status: number;
        setupTimeInMinutes: number;
        operationalWindow: OperationalWindow;
        qualifications?: Qualification[];
    }) {
        if (!params.code) throw new Error('Code cannot be null or empty.');
        if (!params.description) throw new Error('Description cannot be null or empty.');
        if (!params.operationalWindow) throw new Error('Operational window cannot be null.');

        this._type = params.type;
        this._code = params.code;
        this._description = params.description;
        this._status = params.status;
        this._setupTimeInMinutes = params.setupTimeInMinutes;
        this._operationalWindow = params.operationalWindow;
        this._qualifications = params.qualifications ?? [];
    }

    get type(): string { return this._type; }
    get code(): string { return this._code; }
    get description(): string { return this._description; }
    get status(): ResourceStatus { return this._status; }
    get setupTimeInMinutes(): number { return this._setupTimeInMinutes; }
    get operationalWindow(): OperationalWindow { return this._operationalWindow; }
    get qualifications(): Qualification[] { return this._qualifications; }
    get active(): boolean { return this._active; }

    deactivate(): void {
        this._active = false;
        this._status = ResourceStatus.OutOfService;
    }

    updateDescription(description: string): void {
        if (!description) throw new Error('Description cannot be null or empty.');
        this._description = description;
    }

    updateStatus(status: ResourceStatus): void {
        this._status = status;
    }

    updateSetupTime(setupTimeInMinutes: number): void {
        this._setupTimeInMinutes = setupTimeInMinutes;
    }

    updateQualifications(qualifications: Qualification[]): void {
        this._qualifications = qualifications;
    }

    updateOperationalWindow(operationalWindow: OperationalWindow): void {
        if (!operationalWindow) throw new Error('Operational window cannot be null.');
        this._operationalWindow = operationalWindow;
    }

    abstract toDto(): PhysicalResourceDto;
}

export class STSCrane extends PhysicalResource {
    private _liftingCapacity: number;
    private _servingDock: Dock;
    private _containersPerHour: number;

    constructor(params: {
        code: string;
        description: string;
        status: number;
        setupTimeInMinutes: number;
        operationalWindow: OperationalWindow;
        qualifications?: Qualification[];
        liftingCapacity: number;
        servingDock: Dock;
        containersPerHour: number;
    }) {
        super({
            type: 'STS Crane',
            code: params.code,
            description: params.description,
            status: params.status,
            setupTimeInMinutes: params.setupTimeInMinutes,
            operationalWindow: params.operationalWindow,
            qualifications: params.qualifications
        });
        this._liftingCapacity = params.liftingCapacity;
        this._servingDock = params.servingDock;
        this._containersPerHour = params.containersPerHour;
    }

    get liftingCapacity(): number { return this._liftingCapacity; }
    get servingDock(): Dock { return this._servingDock; }
    get containersPerHour(): number { return this._containersPerHour; }

    updateServingDock(dock: Dock): void {
        if (!dock) throw new Error('Serving dock cannot be null.');
        this._servingDock = dock;
    }

    updateLiftingCapacity(liftingCapacity: number): void {
        this._liftingCapacity = liftingCapacity;
    }

    updateContainersPerHour(containersPerHour: number): void {
        this._containersPerHour = containersPerHour;
    }

    toDto(): STSCraneDto {
        return {
            type: 'STS Crane',
            code: this._code,
            description: this._description,
            status: this._status,
            setupTimeInMinutes: this._setupTimeInMinutes,
            operationalWindow: this._operationalWindow,
            qualificationsCodes: this._qualifications.map(q => q.idCode),
            liftingCapacity: this._liftingCapacity,
            servingDockCode: this._servingDock.code,
            containersPerHour: this._containersPerHour
        } as any as STSCraneDto;
    }
}

export class YardCrane extends PhysicalResource {
    private _liftingCapacity: number;
    private _containersPerHour: number;

    constructor(params: {
        code: string;
        description: string;
        status: number;
        setupTimeInMinutes: number;
        operationalWindow: OperationalWindow;
        qualifications?: Qualification[];
        liftingCapacity: number;
        containersPerHour: number;
    }) {
        super({
            type: 'Yard Crane',
            code: params.code,
            description: params.description,
            status: params.status,
            setupTimeInMinutes: params.setupTimeInMinutes,
            operationalWindow: params.operationalWindow,
            qualifications: params.qualifications
        });
        this._liftingCapacity = params.liftingCapacity;
        this._containersPerHour = params.containersPerHour;
    }

    get liftingCapacity(): number { return this._liftingCapacity; }
    get containersPerHour(): number { return this._containersPerHour; }

    updateLiftingCapacity(liftingCapacity: number): void {
        this._liftingCapacity = liftingCapacity;
    }

    updateContainersPerHour(containersPerHour: number): void {
        this._containersPerHour = containersPerHour;
    }

    toDto(): YardCraneDto {
        return {
            type: 'Yard Crane',
            code: this._code,
            description: this._description,
            status: this._status,
            setupTimeInMinutes: this._setupTimeInMinutes,
            operationalWindow: this._operationalWindow,
            qualificationsCodes: this._qualifications.map(q => q.idCode),
            liftingCapacity: this._liftingCapacity,
            containersPerHour: this._containersPerHour
        } as any as YardCraneDto;
    }
}

export class Truck extends PhysicalResource {
    private _maxLoadCapacity: number;
    private _averageSpeed: number;
    private _containersPerTrip: number;

    constructor(params: {
        code: string;
        description: string;
        status: number;
        setupTimeInMinutes: number;
        operationalWindow: OperationalWindow;
        qualifications?: Qualification[];
        maxLoadCapacity: number;
        averageSpeed: number;
        containersPerTrip: number;
    }) {
        super({
            type: 'Truck',
            code: params.code,
            description: params.description,
            status: params.status,
            setupTimeInMinutes: params.setupTimeInMinutes,
            operationalWindow: params.operationalWindow,
            qualifications: params.qualifications
        });
        this._maxLoadCapacity = params.maxLoadCapacity;
        this._averageSpeed = params.averageSpeed;
        this._containersPerTrip = params.containersPerTrip;
    }

    get maxLoadCapacity(): number { return this._maxLoadCapacity; }
    get averageSpeed(): number { return this._averageSpeed; }
    get containersPerTrip(): number { return this._containersPerTrip; }

    updateContainersPerTrip(containersPerTrip: number): void {
        this._containersPerTrip = containersPerTrip;
    }

    updateAverageSpeed(averageSpeed: number): void {
        this._averageSpeed = averageSpeed;
    }

    updateMaxLoadCapacity(maxLoadCapacity: number): void {
        this._maxLoadCapacity = maxLoadCapacity;
    }

    toDto(): TruckDto {
        return {
            type: 'Truck',
            code: this._code,
            description: this._description,
            status: this._status,
            setupTimeInMinutes: this._setupTimeInMinutes,
            operationalWindow: this._operationalWindow,
            qualificationsCodes: this._qualifications.map(q => q.idCode),
            maxLoadCapacity: this._maxLoadCapacity,
            averageSpeed: this._averageSpeed,
            containersPerTrip: this._containersPerTrip
        } as any as TruckDto;
    }
}